File: patterns/architectural/clean-architecture.py
"""
Clean Architecture Pattern - Python

Separates software into layers with clear dependencies:
- Entities: Business objects
- Use Cases: Application-specific business rules
- Interface Adapters: Controllers, presenters, gateways
- Frameworks & Drivers: External tools and frameworks
"""


# Entities (Domain Layer)
class User:
    def __init__(self, id, name, email):
        self.id = id
        self.name = name
        self.email = email


# Use Cases (Application Layer)
class CreateUserUseCase:
    def __init__(self, user_repository):
        self.user_repository = user_repository
    
    def execute(self, name, email):
        if not name or not email:
            raise ValueError("Name and email are required")
        
        user = User(int(__import__('time').time()), name, email)
        return self.user_repository.save(user)


# Interface Adapters (Interface Layer)
class UserRepository:
    def __init__(self):
        self.users = []
    
    def save(self, user):
        self.users.append(user)
        return user
    
    def find_by_id(self, id):
        return next((u for u in self.users if u.id == id), None)


class UserController:
    def __init__(self, create_user_use_case):
        self.create_user_use_case = create_user_use_case
    
    def create_user(self, request):
        try:
            user = self.create_user_use_case.execute(request['name'], request['email'])
            return {'success': True, 'user': user}
        except ValueError as error:
            return {'success': False, 'error': str(error)}


# Example usage
if __name__ == "__main__":
    print("=== Clean Architecture Pattern ===\n")
    
    repository = UserRepository()
    create_user_use_case = CreateUserUseCase(repository)
    controller = UserController(create_user_use_case)
    
    result = controller.create_user({
        'name': 'John Doe',
        'email': 'john@example.com'
    })
    
    print("Result:", result)

File: patterns/architectural/clean-architecture.js
/**
 * Clean Architecture Pattern - JavaScript
 * 
 * Separates software into layers with clear dependencies:
 * - Entities: Business objects
 * - Use Cases: Application-specific business rules
 * - Interface Adapters: Controllers, presenters, gateways
 * - Frameworks & Drivers: External tools and frameworks
 */

// Entities (Domain Layer)
class User {
    constructor(id, name, email) {
        this.id = id;
        this.name = name;
        this.email = email;
    }
}

// Use Cases (Application Layer)
class CreateUserUseCase {
    constructor(userRepository) {
        this.userRepository = userRepository;
    }
    
    execute(name, email) {
        if (!name || !email) {
            throw new Error('Name and email are required');
        }
        
        const user = new User(Date.now(), name, email);
        return this.userRepository.save(user);
    }
}

// Interface Adapters (Interface Layer)
class UserRepository {
    constructor() {
        this.users = [];
    }
    
    save(user) {
        this.users.push(user);
        return user;
    }
    
    findById(id) {
        return this.users.find(u => u.id === id);
    }
}

class UserController {
    constructor(createUserUseCase) {
        this.createUserUseCase = createUserUseCase;
    }
    
    createUser(request) {
        try {
            const user = this.createUserUseCase.execute(request.name, request.email);
            return { success: true, user };
        } catch (error) {
            return { success: false, error: error.message };
        }
    }
}

// Example usage
console.log('=== Clean Architecture Pattern ===\n');

const repository = new UserRepository();
const createUserUseCase = new CreateUserUseCase(repository);
const controller = new UserController(createUserUseCase);

const result = controller.createUser({
    name: 'John Doe',
    email: 'john@example.com'
});

console.log('Result:', result);

module.exports = { User, CreateUserUseCase, UserRepository, UserController };

